File: packages/schema/src/migrations/v0_5_to_v0_6.ts
/**
 * Migration from schema version 0.5.0 to 0.6.0.
 *
 * Converts legacy representations to Grammar symbologyState in one pass:
 *  - parameters.color (flat OL FlatStyle keys) → Grammar directly
 *  - parameters.symbologyState with old render types → Grammar
 *  - HeatmapLayer color array → symbologyState.gradient
 *  - WebGlLayer type → GeoTiffLayer
 *  - layer.filters → grammar layer-level when + whenOp (filters removed)
 *  - flat metadata keys ``annotation_<id>`` → top-level ``annotations``
 */

import {
  categorizedToGrammar,
  graduatedToGrammar,
  singleSymbolToGrammar,
  SymbologyState,
} from '../grammar/grammarConversions';

const ANNOTATION_KEY_PATTERN = /^annotation_(.+)$/;

export function migrate(doc: Record<string, any>): Record<string, any> {
  const layers: Record<string, any> = { ...doc.layers };

  for (const [id, layer] of Object.entries(layers)) {
    const newType = layer.type === 'WebGlLayer' ? 'GeoTiffLayer' : layer.type;
    const params = layer?.parameters;

    if (!params) {
      if (newType !== layer.type) {
        layers[id] = { ...layer, type: newType };
      }
      continue;
    }

    const newParams = { ...params };

    if (layer.type === 'VectorLayer' || layer.type === 'VectorTileLayer') {
      const color = newParams.color;
      if (color && typeof color === 'object' && !Array.isArray(color)) {
        // Flat OL color dict → Grammar (skip intermediate symbologyState format)
        newParams.symbologyState = singleSymbolToGrammar(
          _colorToState(color as Record<string, unknown>),
        );
        delete newParams.color;
      } else {
        // Handle files that already carry old-style symbologyState (defensive)
        const state: SymbologyState | undefined = newParams.symbologyState;
        if (state?.renderType && state.renderType !== 'Grammar') {
          switch (state.renderType) {
            case 'Single Symbol':
              newParams.symbologyState = singleSymbolToGrammar(state);
              break;
            case 'Graduated':
              newParams.symbologyState = graduatedToGrammar(state);
              break;
            case 'Categorized':
              newParams.symbologyState = categorizedToGrammar(state);
              break;
          }
        }
      }
    } else if (layer.type === 'HeatmapLayer') {
      if (Array.isArray(newParams.color)) {
        const state = newParams.symbologyState ?? { renderType: 'Heatmap' };
        if (!state.gradient) {
          newParams.symbologyState = { ...state, gradient: newParams.color };
        }
        delete newParams.color;
      }
    }

    // Convert legacy layer.filters → grammar layer-level when + whenOp.
    const filters = layer.filters;
    if (filters?.appliedFilters?.length) {
      const logicalOp: 'all' | 'any' =
        filters.logicalOp === 'any' ? 'any' : 'all';
      const predicates: Record<string, any>[] = [];

      for (const item of filters.appliedFilters as Record<string, any>[]) {
        if (item.operator === 'between') {
          if (item.betweenMin !== undefined && item.betweenMax !== undefined) {
            predicates.push({
              type: 'between',
              field: item.feature,
              min: item.betweenMin,
              max: item.betweenMax,
            });
          }
        } else if (item.operator === '==') {
          predicates.push({
            type: 'fieldEquals',
            field: item.feature,
            value: item.value,
          });
        } else {
          predicates.push({
            type: 'fieldCompare',
            field: item.feature,
            op: item.operator,
            value: item.value,
          });
        }
      }

      if (predicates.length) {
        const symbologyState = newParams.symbologyState;
        if (
          Array.isArray(symbologyState?.layers) &&
          symbologyState.layers.length > 0
        ) {
          newParams.symbologyState = {
            ...symbologyState,
            layers: symbologyState.layers.map((gl: Record<string, any>) => ({
              ...gl,
              when: [...(gl.when ?? []), ...predicates],
              whenOp: gl.whenOp ?? logicalOp,
            })),
          };
        }
      }
    }

    const { filters: _, ...layerWithoutFilters } = { ...layer, type: newType };
    layers[id] = { ...layerWithoutFilters, parameters: newParams };
  }

  return _migrateAnnotations({ ...doc, layers });
}

function _migrateAnnotations(doc: Record<string, any>): Record<string, any> {
  const metadata = { ...(doc.metadata ?? {}) };
  const annotations: Record<string, any> =
    doc.annotations &&
    typeof doc.annotations === 'object' &&
    !Array.isArray(doc.annotations)
      ? { ...doc.annotations }
      : {};

  const nested = metadata.annotations;
  if (nested && typeof nested === 'object' && !Array.isArray(nested)) {
    Object.assign(annotations, nested);
  }

  for (const [key, value] of Object.entries(metadata)) {
    const match = ANNOTATION_KEY_PATTERN.exec(key);
    if (!match) {
      continue;
    }

    let annotation = value;
    if (typeof annotation === 'string') {
      try {
        annotation = JSON.parse(annotation);
      } catch {
        continue;
      }
    }

    annotations[match[1]] = annotation;
  }

  return {
    ...doc,
    annotations,
    metadata: {},
    presets: {},
  };
}

/** Build a SymbologyState from a flat OL color dict. */
function _colorToState(color: Record<string, unknown>): SymbologyState {
  const state: SymbologyState = { renderType: 'Single Symbol' };

  const fill = _firstSolidColor(
    color['fill-color'],
    color['circle-fill-color'],
  );
  if (fill) {
    state.fillColor = fill;
  }

  const stroke = _firstSolidColor(
    color['stroke-color'],
    color['circle-stroke-color'],
  );
  if (stroke) {
    state.strokeColor = stroke;
  }

  const sw = color['stroke-width'] ?? color['circle-stroke-width'];
  if (typeof sw === 'number') {
    state.strokeWidth = sw;
  }

  const r = color['circle-radius'];
  if (typeof r === 'number') {
    state.radius = r;
  }

  return state;
}

function _firstSolidColor(...candidates: unknown[]): number[] | undefined {
  for (const c of candidates) {
    if (Array.isArray(c) && c.length >= 3 && typeof c[0] === 'number') {
      const [r = 0, g = 0, b = 0, a = 1] = c as number[];
      return [r, g, b, a];
    }
    if (typeof c === 'string') {
      const m = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(c);
      if (m) {
        return [parseInt(m[1], 16), parseInt(m[2], 16), parseInt(m[3], 16), 1];
      }
    }
  }
  return undefined;
}

File: python/jupytergis_core/jupytergis_core/migrations/v0_5_to_v0_6.py
"""Migration from schema version 0.5.0 to 0.6.0.

Converts legacy representations to Grammar symbologyState in one pass:
  - parameters.color (flat OL FlatStyle keys) → Grammar directly
  - parameters.symbologyState with old render types → Grammar
  - HeatmapLayer color array → symbologyState.gradient
  - WebGlLayer type → GeoTiffLayer
"""

import uuid
from typing import Any

_DEFAULT_STROKE_WIDTH = 1.25
_DEFAULT_FILL = [255, 255, 255, 0.4]
_DEFAULT_STROKE = [51, 153, 204, 1]
_DEFAULT_RADIUS = 5
_TRANSPARENT = [0, 0, 0, 0]


def migrate(doc: dict[str, Any]) -> dict[str, Any]:
    layers = dict(doc.get("layers", {}))

    for layer_id, layer in layers.items():
        layer = dict(layer)

        if layer.get("type") == "WebGlLayer":
            layer["type"] = "GeoTiffLayer"

        params = dict(layer.get("parameters", {}))
        layer_type = layer.get("type", "")

        if layer_type in ("VectorLayer", "VectorTileLayer"):
            color = params.get("color")
            if isinstance(color, dict):
                # Flat OL color dict → Grammar (skip intermediate symbologyState)
                params["symbologyState"] = _color_to_grammar(color)
                del params["color"]
            else:
                # Handle files that already carry old-style symbologyState (defensive)
                state = params.get("symbologyState") or {}
                render_type = state.get("renderType")
                if render_type and render_type != "Grammar":
                    grammar = _to_grammar(state)
                    if grammar is not None:
                        params["symbologyState"] = grammar

        elif layer_type == "HeatmapLayer":
            color = params.get("color")
            if isinstance(color, list):
                state = dict(params.get("symbologyState") or {"renderType": "Heatmap"})
                if "gradient" not in state:
                    state["gradient"] = color
                params["symbologyState"] = state
                del params["color"]

        layer["parameters"] = params
        layers[layer_id] = layer

    return {**doc, "layers": layers}


def _color_to_grammar(color: dict[str, Any]) -> dict[str, Any]:
    """Convert a flat OL color dict directly to Grammar (Single Symbol)."""
    fill = _to_rgba(color.get("fill-color") or color.get("circle-fill-color"))
    stroke = _to_rgba(
        color.get("stroke-color") or color.get("circle-stroke-color"),
    )
    stroke_width = color.get("stroke-width") or color.get("circle-stroke-width")
    radius = color.get("circle-radius")

    return _single_symbol(
        fill=fill or _DEFAULT_FILL,
        stroke=stroke or _DEFAULT_STROKE,
        stroke_width=float(stroke_width)
        if isinstance(stroke_width, (int, float))
        else _DEFAULT_STROKE_WIDTH,
        radius=float(radius) if isinstance(radius, (int, float)) else _DEFAULT_RADIUS,
    )


def _to_grammar(state: dict[str, Any]) -> dict[str, Any] | None:
    render_type = state.get("renderType")
    if render_type == "Single Symbol":
        return _single_symbol(
            fill=state.get("fillColor") or _DEFAULT_FILL,
            stroke=state.get("strokeColor") or _DEFAULT_STROKE,
            stroke_width=state.get("strokeWidth")
            if state.get("strokeWidth") is not None
            else _DEFAULT_STROKE_WIDTH,
            radius=state.get("radius")
            if state.get("radius") is not None
            else _DEFAULT_RADIUS,
        )
    if render_type == "Graduated":
        return _graduated(state)
    if render_type == "Categorized":
        return _categorized(state)
    return None


def _single_symbol(
    fill: list,
    stroke: list,
    stroke_width: float,
    radius: float,
) -> dict[str, Any]:
    rule = {
        "id": str(uuid.uuid4()),
        "mappings": [
            {
                "scale": {"scheme": "constant_rgba", "params": {"value": fill}},
                "channels": ["fill-color", "circle-fill-color"],
            },
            {
                "scale": {"scheme": "constant_rgba", "params": {"value": stroke}},
                "channels": ["stroke-color", "circle-stroke-color"],
            },
            {
                "scale": {"scheme": "constant_num", "params": {"value": stroke_width}},
                "channels": ["stroke-width", "circle-stroke-width"],
            },
            {
                "scale": {"scheme": "constant_num", "params": {"value": radius}},
                "channels": ["circle-radius"],
            },
        ],
    }
    return {
        "renderType": "Grammar",
        "layers": [{"id": str(uuid.uuid4()), "rules": [rule]}],
    }


def _graduated(state: dict[str, Any]) -> dict[str, Any]:
    field = state.get("value")
    fallback = state.get("fallbackColor") or _TRANSPARENT
    stroke = state.get("strokeColor") or _DEFAULT_STROKE
    stroke_width = (
        state.get("strokeWidth")
        if state.get("strokeWidth") is not None
        else _DEFAULT_STROKE_WIDTH
    )
    radius = state.get("radius") if state.get("radius") is not None else _DEFAULT_RADIUS

    color_ramp_params: dict[str, Any] = {
        "name": state.get("colorRamp") or "viridis",
        "nShades": state.get("nClasses") or 9,
        "mode": state.get("mode") or "equal interval",
        "reverse": state.get("reverseRamp") or False,
        "fallback": fallback,
    }
    vmin = state.get("vmin")
    vmax = state.get("vmax")
    if vmin is not None and vmax is not None:
        color_ramp_params["domain"] = [vmin, vmax]

    color_ramp_scale = {"scheme": "colorRamp", "params": color_ramp_params}

    fill_channels = (
        ["fill-color", "stroke-color", "circle-fill-color", "circle-stroke-color"]
        if state.get("strokeFollowsFill")
        else ["fill-color", "circle-fill-color"]
    )

    mappings: list[dict[str, Any]] = [
        {"scale": color_ramp_scale, "channels": fill_channels},
    ]
    if not state.get("strokeFollowsFill"):
        mappings.append(
            {
                "scale": {"scheme": "constant_rgba", "params": {"value": stroke}},
                "channels": ["stroke-color", "circle-stroke-color"],
            },
        )
    mappings += [
        {
            "scale": {"scheme": "constant_num", "params": {"value": stroke_width}},
            "channels": ["stroke-width", "circle-stroke-width"],
        },
        {
            "scale": {"scheme": "constant_num", "params": {"value": radius}},
            "channels": ["circle-radius"],
        },
    ]

    rule: dict[str, Any] = {"id": str(uuid.uuid4()), "mappings": mappings}
    if field:
        rule["fields"] = [field]

    return {
        "renderType": "Grammar",
        "layers": [{"id": str(uuid.uuid4()), "rules": [rule]}],
    }


def _categorized(state: dict[str, Any]) -> dict[str, Any]:
    field = state.get("value")
    fallback = state.get("fallbackColor") or _TRANSPARENT
    stroke = state.get("strokeColor") or _DEFAULT_STROKE
    stroke_width = (
        state.get("strokeWidth")
        if state.get("strokeWidth") is not None
        else _DEFAULT_STROKE_WIDTH
    )
    radius = state.get("radius") if state.get("radius") is not None else _DEFAULT_RADIUS

    categorical_params: dict[str, Any] = {
        "colorRamp": state.get("colorRamp") or "viridis",
        "reverse": state.get("reverseRamp") or False,
        "fallback": fallback,
    }
    n_classes = state.get("nClasses")
    if n_classes is not None:
        categorical_params["nShades"] = n_classes

    categorical_scale = {"scheme": "categorical", "params": categorical_params}

    fill_channels = (
        ["fill-color", "stroke-color", "circle-fill-color", "circle-stroke-color"]
        if state.get("strokeFollowsFill")
        else ["fill-color", "circle-fill-color"]
    )

    mappings: list[dict[str, Any]] = [
        {"scale": categorical_scale, "channels": fill_channels},
    ]
    if not state.get("strokeFollowsFill"):
        mappings.append(
            {
                "scale": {"scheme": "constant_rgba", "params": {"value": stroke}},
                "channels": ["stroke-color", "circle-stroke-color"],
            },
        )
    mappings += [
        {
            "scale": {"scheme": "constant_num", "params": {"value": stroke_width}},
            "channels": ["stroke-width", "circle-stroke-width"],
        },
        {
            "scale": {"scheme": "constant_num", "params": {"value": radius}},
            "channels": ["circle-radius"],
        },
    ]

    rule: dict[str, Any] = {"id": str(uuid.uuid4()), "mappings": mappings}
    if field:
        rule["fields"] = [field]

    return {
        "renderType": "Grammar",
        "layers": [{"id": str(uuid.uuid4()), "rules": [rule]}],
    }


def _to_rgba(value: Any) -> list[float] | None:
    if (
        isinstance(value, (list, tuple))
        and value
        and isinstance(value[0], (int, float))
    ):
        rgba = list(value) + [1.0] * (4 - len(value))
        return [float(c) for c in rgba[:4]]
    if isinstance(value, str):
        s = value.lstrip("#")
        if len(s) == 6 and all(c in "0123456789abcdefABCDEF" for c in s):
            return [
                float(int(s[0:2], 16)),
                float(int(s[2:4], 16)),
                float(int(s[4:6], 16)),
                1.0,
            ]
    return None
